### PR TITLE
merge serena_simple to master after review

### PR DESCRIPTION
**Pull request by serena049**
review ETL codes and merge to master if ready

### CODE REVIEW

File: src/forecast/prediction.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+from sklearn.ensemble import RandomForestClassifier
+import pathlib as pl
+from sklearn.metrics import roc_auc_score, roc_curve
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class Forecast:
+    def __init__(self, df_fcst, model):
+        self.df_fcst = df_fcst
```
Comment (CallMeK) [inline src/forecast/prediction.py:11]: abstract variable name without documentation, what is fcst?

File: src/forecast/prediction.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        self.model = model
+
+    def forecast(self):
+        x_forecast = self.df_fcst.values
```
Comment (CallMeK) [inline src/forecast/prediction.py:19]: probably need to document the assumed format (columns) for df_fcst.

File: src/forecast/prediction.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from sklearn.svm import SVC
+from sklearn.ensemble import GradientBoostingClassifier
+from sklearn.ensemble import RandomForestClassifier
+import pathlib as pl
+from sklearn.metrics import roc_auc_score, roc_curve
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class Forecast:
```
Comment (CallMeK) [inline src/forecast/prediction.py]: If it is going to be a class, probably better to be called Forecaster.

File: src/main.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    # step 1: ETL
+    # read in raw data
+    input_data_path = pl.Path(__file__).resolve().parents[1].joinpath('data/input/')  # the parent directory of
+    check_data = etl.Check(input_data_path)
+    df_all_files = check_data.list_all_files()
+    check_data.check_path(df_all_files)
+    check_data.convert(df_all_files)
+    df_raw = check_data.load(df_all_files)
+    check_data.check_size(df_raw)
+
```
Comment (CallMeK) [inline src/main.py]: Your main file should only contains simple function calls, such as, load_data_and_check, transform_data, etc.

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    df_rv_na = df_transformation.remove_missing_value()
+    df_encode = df_transformation.preprocess(df_rv_na)
+
+    # Step 2: Fit models
+    # train test split
+    df_split = fitmodel.AsNumpy(df_encode)
+    train_x, train_y, test_x, test_y, cols = df_split.train_test_split()
+
+    # run a logistic regression model
+    logit = LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
```
Comment (CallMeK) [inline src/main.py]: You defined a model, you didn't "run" it. Your comment above is inaccurate

File: src/main.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                                        criterion='friedman_mse', min_samples_split=2, min_samples_leaf=1,
+                                        min_weight_fraction_leaf=0.0, max_depth=3, min_impurity_decrease=0.0,
+                                        min_impurity_split=None, init=None, random_state=None, max_features=None,
+                                        verbose=0, max_leaf_nodes=None, warm_start=False, presort='deprecated',
+                                        validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)
+
+    # fit and compare models
+    list_algorithm_names = ['logit', 'gbtree', 'rf']
+    list_algorithms = [logit, gbtree, randomforest]
+    list_feature_imp_cols = ["coefficients", "features", "features"]
```
Comment (CallMeK) [inline src/main.py]: abstract variable name, why there are two "features"?

File: src/main.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        models[algorithm_name] = model_fit
+        # calculate feature importance
+        feature_importance = model.feature_importance(model_fit)
+        feature_importance.to_excel(writer_fi, sheet_name=algorithm_name)
+        # evaluate model performance
+        evl, fpr, tpr, auc = model.evaluation(predictions, probabilities)
+        pd.DataFrame([evl]).to_excel(writer_evl, sheet_name=algorithm_name)
+        # ROC curve
+        plt_path = output_data_path.joinpath(algorithm_name + '.png')
+        model.roc_curve(fpr, tpr, auc, plt_path, algorithm_name)
```
Comment (CallMeK) [inline src/main.py]: details, shouldn't not be in main.

File: src/model/fitmodel.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+from sklearn.svm import SVC
+from sklearn.ensemble import GradientBoostingClassifier
+from sklearn.ensemble import RandomForestClassifier
+import pathlib as pl
+from sklearn.metrics import roc_auc_score, roc_curve
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class AsNumpy:
```
Comment (CallMeK) [inline src/model/fitmodel.py]: I guess you copy from somewhere, but I guess it is better called NumpyWrapper, or something similar.

File: src/model/fitmodel.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    def evaluation(self, predictions, probabilities):
+        fpr, tpr, thresholds = roc_curve(self.test_y, probabilities[:, 1])
+        auc = roc_auc_score(self.test_y, predictions)
+
+        evl = {}
+        evl["accuracy"] = accuracy_score(self.test_y, predictions)
+        evl["classification report"] = classification_report(self.test_y, predictions)
+        evl["confusion materix"] = confusion_matrix(self.test_y, predictions)
+        print("evalution done!")
+        return evl, fpr, tpr, auc
```
Comment (CallMeK) [inline src/model/fitmodel.py:82]: optional, probably better to output a dictionary with metric name.

File: src/preprocess/etl.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+    def list_all_files(self) -> pd.DataFrame:
+        """
+        go through all the files in the data_analysis directory and build a DataFrame with the file name,
+        parent path and modified time.
+        Returns: a pandas df
+        """
+
+        all_files = []
+        for i in self.data_path.rglob('*.*'):
```
Comment (CallMeK) [inline src/preprocess/etl.py:32]: why you need to list all files?

File: src/preprocess/etl.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            data.columns = map(str.lower, data.columns)
+
+            return data
+        else:
+            raise ValueError("file not exists!")
+
+    @staticmethod
+    def check_size(data: pd.DataFrame, n_cols: int = 21):
+        """
+        Check # of columns and if the data is empty
```
Comment (CallMeK) [inline src/preprocess/etl.py:141]: if empty, do what?

File: src/preprocess/etl.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+from sklearn.preprocessing import StandardScaler
+from sklearn.preprocessing import LabelEncoder
+
+
+__author__ = "Wei (Serena) Zou"
+__copyright__ = "Wei (Serena) Zou"
+__license__ = "mit"
+
+
+class Check:
```
Comment (CallMeK) [inline src/preprocess/etl.py]: class name too general, and it feels like a function, not a class

File: src/forecast/prediction.py (hunk 3 above)
Comment (serena049) [reply src/forecast/prediction.py]: changed accordingly

File: src/model/fitmodel.py (hunk 8 above)
Comment (serena049) [reply src/model/fitmodel.py]: updated class name

File: src/model/fitmodel.py (hunk 9 above)
Comment (serena049) [reply src/model/fitmodel.py:82]: Do you mean sth like {'train_x': train_x, 'train_y': train_y, ...}? Why is dic better than np.arrays? 
One output vs. Four?

File: src/preprocess/etl.py (hunk 10 above)
Comment (serena049) [reply src/preprocess/etl.py:32]: this will be used as an input for function "check_path" (if the input file exits) and "convert" (find the most recent data if the input folder saves all the versions of historical data)

File: src/preprocess/etl.py (hunk 11 above)
Comment (serena049) [reply src/preprocess/etl.py:141]: good point to separate empty vs. other value errors, added an "if" condition for empty data

File: src/main.py (hunk 5 above)
Comment (serena049) [reply src/main.py]: yes and comments updated

File: src/main.py (hunk 6 above)
Comment (serena049) [reply src/main.py]: this indicates if the corresponding algorithm will output feature importance as "coefficients" or "feature importance scores", since both 'gbtree' and 'rf' are tree-based models, they will both output "feature importance scores". Refer to Model.feature_importance for details; Any suggestion on making it more clear?

File: src/main.py (hunk 4 above)
Comment (serena049) [reply src/main.py]: Need suggestions on how to realize this: do I need to generate a "sub-main.py" for each of the module and then just call them in the main.py?

File: src/main.py (hunk 7 above)
Comment (serena049) [reply src/main.py]: similar to a previous comment on what to include in main.py, should I include this in a separate .py just for model outputs?

File: src/preprocess/etl.py [hunk 13]
```diff
@@ -212,18 +212,20 @@ def preprocess(df, target_col: str = 'churn'):
 
 
 class DataPreprocess:
-    def __init__(self, data_path_parent_level: int = 2):
+    def __init__(self, contains_str: str = 'dataset', n_cols: int = 21, data_path_parent_level: int = 2):
         self.data_path_parent_level = data_path_parent_level
```
Comment (CallMeK) [inline src/preprocess/etl.py:221]: what does this contains_str mean? There should be documentation.

File: src/preprocess/etl.py [hunk 14]
```diff
@@ -212,18 +212,20 @@ def preprocess(df, target_col: str = 'churn'):
...
+        self.contains_str = contains_str
+        self.n_cols = n_cols
 
     def load_and_encode_data(self):
         # the parent directory of input data path
         input_data_path = pl.Path(__file__).resolve().parents[self.data_path_parent_level].joinpath('data/input/')
         check_data = CheckRawInputs(input_data_path)
```
Comment (CallMeK) [inline src/preprocess/etl.py]: I feel if it is a class, you should call it RawInputsChecker or RawInputsProcessor, etc

File: src/model/fitmodel.py [hunk 15]
```diff
@@ -158,16 +158,16 @@ class FitModel:
 
     all_models = zip(list_of_models_names, list_algorithms, list_feature_imp_cols)
 
-    def __init__(self, all_models, df_encode: pd.DataFrame, list_of_models: list = None, data_path_parent_level:
-    int = 1):
+    def __init__(self, df_encode: pd.DataFrame, list_of_models: list = None, data_path_parent_level:
```
Comment (CallMeK) [inline src/model/fitmodel.py]: For the type hint, can you use List[str]? it is important to denote what is inside the list.

File: src/model/fitmodel.py [hunk 16]
```diff
@@ -158,16 +158,16 @@ class FitModel:
...
+    int = 2):
         """
 
         :type df_encode: object
         """
         if list_of_models is None:
             list_of_models = ['logit', 'gbtree', 'rf', 'gnb', 'svc_lin']
         # select the models user want to test
-        self.selected_models = [item for item in all_models if item[0] in list_of_models]
+        self.selected_models = [item for item in FitModel.all_models if item[0] in list_of_models]
```
Comment (CallMeK) [inline src/model/fitmodel.py]: nit: it is usually better to use a function to return the internal variables.

File: src/forecast/prediction.py [hunk 17]
```diff
@@ -33,14 +23,23 @@ def forecast(self):
...
 
-if __name__ == '__main__':
-    print("done!")
-
-
-
-
+class FcstAndOutputs:
+    def __init__(self, df_fcst_encode: pd.DataFrame, data_path_parent_level: int = 2, selected_model=None):
+        if selected_model is None:
```
Comment (CallMeK) [inline src/forecast/prediction.py]: Google's practice is to just use " if not selected_model"

**Review by CallMeK — COMMENTED** [review]
Much better!!

File: src/preprocess/etl.py (hunk 14 above)
Comment (serena049) [reply src/preprocess/etl.py]: updated!

File: src/preprocess/etl.py (hunk 13 above)
Comment (serena049) [reply src/preprocess/etl.py:221]: added documentation

File: src/model/fitmodel.py (hunk 15 above)
Comment (serena049) [reply src/model/fitmodel.py]: this is a good point, updated!

File: src/forecast/prediction.py (hunk 17 above)
Comment (serena049) [reply src/forecast/prediction.py]: updated!

File: src/model/fitmodel.py (hunk 16 above)
Comment (serena049) [reply src/model/fitmodel.py]: yes and updated
